pkg/search/postgres/storecache.go

```diff
@@ -0,0 +1,329 @@
+	defer c.cacheLock.Unlock()
+	if dbErr != nil {
+		c.resetCacheNoLock()
+		return dbErr
+	}
+	delete(c.cache, id)
+	return nil
+}
+
+// DeleteMany removes the objects associated to the specified IDs from the store.
+func (c *CachedStore[T, PT]) DeleteMany(ctx context.Context, identifiers []string) error {
+	dbErr := c.underlyingStore.DeleteMany(ctx, identifiers)
+	c.cacheLock.Lock()
+	defer c.cacheLock.RUnlock()
+	if dbErr != nil {
+		c.resetCacheNoLock()
+		return dbErr
+	}
+	for _, id := range identifiers {
+		delete(c.cache, id)
+	}
+	return nil
+}
+
+// Exists tells whether the ID exists in the store.
+func (c *CachedStore[T, PT]) Exists(ctx context.Context, id string) (bool, error) {
+	c.cacheLock.RLock()
+	defer c.cacheLock.RUnlock()
+	obj, found := c.cache[id]
+	if !found {
+		return false, nil
+	}
+	return c.isReadAllowed(ctx, obj), nil
+}
+
+// Count returns the number of objects in the store.
+func (c *CachedStore[T, PT]) Count(ctx context.Context) (int, error) {
+	c.cacheLock.RLock()
+	defer c.cacheLock.RUnlock()
+	count := 0
+	walkErr := c.walkCacheNoLock(ctx, func(obj PT) error {
+		count++
+		return nil
+	})
+	if walkErr != nil {
+		return 0, walkErr
+	}
+	return count, nil
+}
+
+// Get returns the object, if it exists from the store.
+func (c *CachedStore[T, PT]) Get(ctx context.Context, id string) (PT, bool, error) {
+	c.cacheLock.RLock()
+	defer c.cacheLock.RUnlock()
+	obj, found := c.cache[id]
+	if !found {
+		return nil, false, nil
+	}
+	if !c.isReadAllowed(ctx, obj) {
+		return nil, false, nil
+	}
+	return obj, true, nil
+}
+
+// GetMany returns the objects specified by the IDs from the store as well as the index in the missing indices slice.
+func (c *CachedStore[T, PT]) GetMany(ctx context.Context, identifiers []string) ([]PT, []int, error) {
+	if len(identifiers) == 0 {
+		return nil, nil, nil
+	}
+	c.cacheLock.RLock()
+	defer c.cacheLock.RUnlock()
+	results := make([]PT, 0, len(identifiers))
+	misses := make([]int, 0, len(identifiers))
+	for ix, id := range identifiers {
+		obj, found := c.cache[id]
+		if !found {
+			misses = append(misses, ix)
+			continue
+		}
+		if !c.isReadAllowed(ctx, obj) {
+			misses = append(misses, ix)
+			continue
+		}
+		results = append(results, obj)
+	}
+	return results, misses, nil
+}
+
+// Walk iterates over all the objects in the store and applies the closure.
+func (c *CachedStore[T, PT]) Walk(ctx context.Context, fn func(obj PT) error) error {
+	c.cacheLock.RLock()
+	defer c.cacheLock.RUnlock()
+	return c.walkCacheNoLock(ctx, fn)
+}
+
+// GetByQuery returns the objects from the store matching the query.
+func (c *CachedStore[T, PT]) GetByQuery(ctx context.Context, query *v1.Query) ([]*T, error) {
+	return c.underlyingStore.GetByQuery(ctx, query)
+}
+
+// DeleteByQuery removes the objects from the store based on the passed query.
+func (c *CachedStore[T, PT]) DeleteByQuery(ctx context.Context, query *v1.Query) error {
+	dbErr := c.underlyingStore.DeleteByQuery(ctx, query)
```

Comment:
I'll use GetByQuery from the underlying store and delete by identifiers. This might pull a lot of data, but the function does not seem to be used from many places, so the impact might still be contained.

Reply:
This was split out of the current PR and moved to #8975